packages/miniflare/src/workers/d1/dumpSql.ts

```diff
@@ -0,0 +1,152 @@
+
+		if (table === "sqlite_sequence") {
+			if (!noSchema) yield `DELETE FROM sqlite_sequence;`;
+		} else if (table.startsWith(`sqlite_stat`)) {
+			// This feels like it should really appear _after_ the contents of sqlite_stat[1,4] but I'm choosing
+			// to match SQLite's dump output exactly so writing it immediately like they do.
+			if (!noSchema) yield `ANALYZE sqlite_schema;`;
+		} else if (sql.startsWith(`CREATE VIRTUAL TABLE`)) {
+			throw new Error(
+				`D1 Export error: cannot export databases with Virtual Tables (fts5)`
+			);
+		} else if (table.startsWith("_cf_") || table.startsWith("sqlite_")) {
+			continue;
+		} else {
+			// SQLite dump has an extremely weird behaviour where, if the table was explicitly
+			// quoted i.e. "Table", then in the dump it has `IF NOT EXISTS` injected. I don't understand
+			// why, but on the off chance there's a good reason to I am following suit.
+			if (sql.match(/CREATE TABLE ['"].*/)) {
+				if (!noSchema) yield `CREATE TABLE IF NOT EXISTS ${sql.substring(13)};`;
+			} else {
+				if (!noSchema) yield `${sql};`;
+			}
+		}
+
+		if (noData) continue;
+		const columns_cursor = db.exec(`PRAGMA table_info="${table}"`);
+		const columns = Array.from(columns_cursor);
+		const select = `SELECT ${columns.map((c) => c.name).join(", ")}
+                            FROM "${table}";`;
+		const rows_cursor = db.exec(select);
+		for (const dataRow of rows_cursor.raw()) {
+			const formattedCells = dataRow.map((cell: unknown, i: number) => {
+				const colType = columns[i].type;
+				const cellType = typeof cell;
+				if (cell === null) {
+					return "NULL";
+				} else if (colType === "INTEGER" || cellType === "number") {
+					return cell;
+				} else if (colType === "TEXT" || cellType === "string") {
+					// Reproduce `output_quoted_escaped_string` from https://github.com/sqlite/sqlite/blob/master/src/shell.c.in#L1799-L1862
+					let lfs = false;
+					let crs = false;
+					// First escape any ' as ''
+					let output_string = `'${(cell as string).replace(
+						/'|(\n)|(\r)/g,
+						(_: string, lf: string, cr: string) => {
+							// While also tracking whether the string contains \r or \n
+							if (lf) {
+								lfs = true;
+								return `\\n`;
+							}
+							if (cr) {
+								crs = true;
+								return `\\r`;
+							}
+							return `''`;
+						}
+					)}'`;
+					if (crs) output_string = `replace(${output_string},'\\r',char(13))`;
+					if (lfs) output_string = `replace(${output_string},'\\n',char(10))`;
+
+					return output_string;
+				} else if (cell instanceof ArrayBuffer) {
+					return `X'${Array.prototype.map
+						.call(new Uint8Array(cell), (b) => b.toString(16))
+						.join("")}'`;
+				} else {
+					console.log({ colType, cellType, cell, column: columns[i] });
+					return "ERROR";
+				}
+			});
+
+			yield `INSERT INTO ${sqliteQuote(table)} VALUES(${formattedCells.join(",")});`;
+		}
+	}
+
+	if (!noSchema) {
```

Comment:
Does it matter that this query appears first in shell.c.in?

Reply:
Does it? This is supposed to match up with https://github.com/sqlite/sqlite/blob/105c20648e1b05839fd0638686b95f2e3998abcb/src/shell.c.in#L8472, i.e. directly after `run_schema_dump_query` (which dumps schema & data), it checks `SHFLG_DumpDataOnly)` is not set before exporting all `('index','trigger','view')` rows of `sqlite_schema`

Have I got that wrong?

Reply:
No, you're correct.